File: Nebokrai/server/rust/src/billingserver/appbilling/playerfillmgr.rs
//! Владелец optional `CPlayerFillMgr` исторического BillingServer.
//!
//! Контракт выборки, broadcast, удаления и worker lifecycle подтверждён точной
//! парой BillingServer EXE/PDB.
//!
//! Один проход получает не более 50 строк `TBL_NeedUpdate` в DB-порядке. Для
//! каждой строки он вызывает фактический `CRsPlayerAccount::GetUserPoint`,
//! строит broadcast `0xFF004 + Account + result [+ point при result == 0]` и
//! только после всего snapshot удаляет из таблицы все прочитанные ID одним
//! вызовом. Ошибка отдельной отправки не отменяет последующие ответы или delete;
//! DB-fallback `GetUserPoint = -2` остаётся у его собственного владельца.
//!
//! Worker один раз выполняет технический `InitConn`, затем проверяет общий
//! `g_bGameThreadExit` перед каждым проходом и безусловно выдерживает `5000 ms`
//! после него. `End` только ждёт завершение: отдельного stop-сообщения эта ветвь
//! `CGame::Release` не посылала. Windows `PeekMessage(0x66A)` не имеет внешнего
//! sender в достигнутом lifecycle и не получает пустой Linux thread-message
//! аналог; фактическую остановку сохраняет общий atomic-флаг.
//!
//! `JoinHandle` заменяет `_beginthreadex/WaitForSingleObject/CloseHandle`,
//! Tiberius owners — COM apartment и статический ADO config, а `CString`, `Vec`
//! и `Drop` — C-string/STL/SEH cleanup. Recordset/ADO, iostream/STL internals,
//! deleting thunks, `$L/$E` cleanup и compiler catch-блоки удалены.

use std::collections::VecDeque;
use std::error::Error;
use std::ffi::CString;
use std::fmt;
use std::io;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::thread::{self, JoinHandle};
use std::time::Duration;

use parking_lot::Mutex;

use crate::dbaccess::dbbilling::rsplayeraccount::{
    BillingDatabaseSettings, RsPlayerAccountInitializationError, RsPlayerAccountNotice,
    RsPlayerAccountOwner, TiberiusRsPlayerAccount,
};
use crate::dbaccess::dbbilling::rsplayerfillmgr::{
    PlayerFillDatabaseSettings, RsPlayerFillInitializationError, RsPlayerFillNotice,
    RsPlayerFillOwner, TiberiusRsPlayerFillMgr,
};
use crate::nets::netbilling::message::CMessage;
use crate::nets::servers::ServerCommandHandle;

const PLAYER_FILL_RESPONSE: i32 = 0x000F_F004;
const PLAYER_FILL_CADENCE: Duration = Duration::from_millis(5_000);

/// DB-owner, который не смог начать работу внутри уже созданного thread.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) enum PlayerFillWorkerOwner {
    FillTable,
    PlayerAccount,
}

/// Operator-visible событие optional PlayerFill worker.
#[derive(Debug)]
pub(crate) enum PlayerFillNotice {
    FillDatabase(RsPlayerFillNotice),
    AccountDatabase(RsPlayerAccountNotice),
    WorkerUnavailable {
        owner: PlayerFillWorkerOwner,
        reason: String,
    },
    WorkerPanicked,
}

/// Точная фатальная граница `Start`: ОС не создала worker thread.
#[derive(Debug)]
pub(crate) struct StartPlayerFillError(io::Error);

impl fmt::Display for StartPlayerFillError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "не создан PlayerFill worker: {}", self.0)
    }
}

impl Error for StartPlayerFillError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        Some(&self.0)
    }
}

/// Общая static-часть исходного `CPlayerFillMgr`.
#[derive(Default)]
pub(crate) struct CPlayerFillMgr {
    notices: Mutex<VecDeque<PlayerFillNotice>>,
}

impl CPlayerFillMgr {
    /// Выполняет один fetch/send/delete проход исходного `Run`.
    pub(crate) fn run(
        &self,
        fill_database: &mut dyn RsPlayerFillOwner,
        account_database: &mut dyn RsPlayerAccountOwner,
        sender: &ServerCommandHandle,
    ) -> bool {
        let entries = fill_database.get_player_fill_log();
        self.collect_fill_notices(fill_database);

        for entry in &entries {
            let outcome = account_database.get_user_point(&entry.player_account);
            self.collect_account_notices(account_database);

            let mut response = CMessage::new(PLAYER_FILL_RESPONSE);
            let account = legacy_c_string(&entry.player_account);
            response.base_mut().add_str(Some(&account));
            response.base_mut().add_long(outcome.result);
            if outcome.result == 0 {
                response.base_mut().add_long(outcome.point);
            }
            let _ = response.send_to_all_gs(sender);
        }

        if !entries.is_empty() {
            fill_database.delete_player_fill_log(&entries);
            self.collect_fill_notices(fill_database);
        }
        true
    }

    /// Забирает следующее DB/worker-событие без credential/account values.
    pub(crate) fn pop_notice(&self) -> Option<PlayerFillNotice> {
        self.notices.lock().pop_front()
    }

    fn collect_fill_notices(&self, database: &mut dyn RsPlayerFillOwner) {
        while let Some(notice) = database.pop_notice() {
            self.push_notice(PlayerFillNotice::FillDatabase(notice));
        }
    }

    fn collect_account_notices(&self, database: &mut dyn RsPlayerAccountOwner) {
        while let Some(notice) = database.pop_notice() {
            self.push_notice(PlayerFillNotice::AccountDatabase(notice));
        }
    }

    fn push_notice(&self, notice: PlayerFillNotice) {
        self.notices.lock().push_back(notice);
    }
}

/// Owned Linux lifecycle единственного optional PlayerFill thread.
pub(crate) struct PlayerFillRuntime {
    manager: Arc<CPlayerFillMgr>,
    fill_database_settings: PlayerFillDatabaseSettings,
    account_database_settings: BillingDatabaseSettings,
    sender: ServerCommandHandle,
    game_thread_exit: Arc<AtomicBool>,
    worker: Option<JoinHandle<()>>,
}

impl PlayerFillRuntime {
    /// Связывает DB settings, broadcast-owner и общий game-thread exit.
    pub(crate) fn new(
        manager: Arc<CPlayerFillMgr>,
        fill_database_settings: PlayerFillDatabaseSettings,
        account_database_settings: BillingDatabaseSettings,
        sender: ServerCommandHandle,
        game_thread_exit: Arc<AtomicBool>,
    ) -> Self {
        Self {
            manager,
            fill_database_settings,
            account_database_settings,
            sender,
            game_thread_exit,
            worker: None,
        }
    }

    /// Создаёт единственный worker; только ошибка thread spawn является `false`.
    pub(crate) fn start(&mut self) -> Result<(), StartPlayerFillError> {
        let manager = Arc::clone(&self.manager);
        let fill_settings = self.fill_database_settings.clone();
        let account_settings = self.account_database_settings.clone();
        let sender = self.sender.clone();
        let game_thread_exit = Arc::clone(&self.game_thread_exit);
        let worker = thread::Builder::new()
            .name("billing-player-fill".to_owned())
            .spawn(move || {
                let mut fill_database = match TiberiusRsPlayerFillMgr::new(fill_settings) {
                    Ok(database) => database,
                    Err(error) => {
                        manager.push_notice(worker_unavailable_fill(error));
                        return;
                    }
                };
                let mut account_database = match TiberiusRsPlayerAccount::new(account_settings) {
                    Ok(database) => database,
                    Err(error) => {
                        manager.push_notice(worker_unavailable_account(error));
                        return;
                    }
                };

                while !game_thread_exit.load(Ordering::Acquire) {
                    manager.run(&mut fill_database, &mut account_database, &sender);
                    thread::sleep(PLAYER_FILL_CADENCE);
                }
            })
            .map_err(StartPlayerFillError)?;
        self.worker = Some(worker);
        Ok(())
    }

    /// Ждёт thread без отдельного stop, как исходный `End`.
    pub(crate) fn end(&mut self) -> bool {
        if let Some(worker) = self.worker.take()
            && worker.join().is_err()
        {
            self.manager.push_notice(PlayerFillNotice::WorkerPanicked);
        }
        true
    }
}

fn legacy_c_string(bytes: &[u8]) -> CString {
    let visible = bytes
        .iter()
        .position(|byte| *byte == 0)
        .map_or(bytes, |end| &bytes[..end]);
    CString::new(visible).expect("slice обрезан по первому NUL")
}

fn worker_unavailable_fill(error: RsPlayerFillInitializationError) -> PlayerFillNotice {
    PlayerFillNotice::WorkerUnavailable {
        owner: PlayerFillWorkerOwner::FillTable,
        reason: error.to_string(),
    }
}

fn worker_unavailable_account(error: RsPlayerAccountInitializationError) -> PlayerFillNotice {
    PlayerFillNotice::WorkerUnavailable {
        owner: PlayerFillWorkerOwner::PlayerAccount,
        reason: error.to_string(),
    }
}

File: server/cpp/src/billing/appbilling/playerfillmgr.cpp
#include "playerfillmgr.h"

#include "../../nets/netbilling/message.h"

#include <spdlog/spdlog.h>

#include <algorithm>
#include <chrono>
#include <string>
#include <utility>

namespace Billing
{
namespace
{
constexpr std::int32_t kPlayerFillResponse = 0x000F'F004;
constexpr auto kCadence = std::chrono::seconds(5);

std::string VisibleString(std::span<const std::uint8_t> bytes)
{
    const auto end = std::find(bytes.begin(), bytes.end(), std::uint8_t{});
    return {bytes.begin(), end};
}
}

CPlayerFillMgr::CPlayerFillMgr(ServerCommandHandle sender,
                               RsPlayerFillFactory fillFactory,
                               RsPlayerAccountFactory accountFactory)
    : m_Sender(std::move(sender)),
      m_FillFactory(std::move(fillFactory)),
      m_AccountFactory(std::move(accountFactory))
{
}

CPlayerFillMgr::~CPlayerFillMgr()
{
    RequestStop();
    Join();
}

bool CPlayerFillMgr::Start()
{
    if (m_Worker.joinable() || !m_FillFactory || !m_AccountFactory) return false;
    m_StopRequested.store(false, std::memory_order_release);
    try {
        m_Worker = std::thread(&CPlayerFillMgr::Worker, this);
    } catch (const std::exception& error) {
        spdlog::error("BillingServer: не создан PlayerFill-worker: {}", error.what());
        return false;
    }
    return true;
}

void CPlayerFillMgr::RequestStop() noexcept
{
    m_StopRequested.store(true, std::memory_order_release);
}

void CPlayerFillMgr::Join() noexcept
{
    if (m_Worker.joinable()) m_Worker.join();
}

void CPlayerFillMgr::Worker()
{
    std::unique_ptr<IRsPlayerFill> fillDatabase = m_FillFactory();
    std::unique_ptr<IRsPlayerAccount> accountDatabase = m_AccountFactory();
    if (!fillDatabase || !accountDatabase) {
        spdlog::error("BillingServer: фабрика не создала PlayerFill DB-owner");
        return;
    }

    while (!m_StopRequested.load(std::memory_order_acquire)) {
        const std::vector<PlayerFillInfo> entries = fillDatabase->GetPlayerFillLog();
        for (const PlayerFillInfo& entry : entries) {
            const UserPointOutcome outcome =
                accountDatabase->GetUserPoint(entry.playerAccount);
            BillingNet::CMessage response(kPlayerFillResponse);
            const std::string account = VisibleString(entry.playerAccount);
            response.Base().Add(account.c_str());
            response.Base().Add(outcome.result);
            if (outcome.result == 0) response.Base().Add(outcome.point);
            static_cast<void>(response.SendToAllGS(m_Sender));
        }
        if (!entries.empty()) {
            static_cast<void>(fillDatabase->DeletePlayerFillLog(entries));
        }

        const auto deadline = std::chrono::steady_clock::now() + kCadence;
        while (!m_StopRequested.load(std::memory_order_acquire) &&
               std::chrono::steady_clock::now() < deadline) {
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
        }
    }
}
}
